### PR TITLE
Icu 12567 add new field credential mapping overrides

### PR DESCRIPTION
**Pull request by DhariniJeeva**
:tickets: [Jira ticket](https://hashicorp.atlassian.net/browse/JIRA_TICKET_NUMBER)

## Screenshots (if appropriate):

 

https://github.com/hashicorp/boundary-ui/assets/15043878/4be505c9-b306-429a-99f9-ea228d7adf49

 
 # How to test 

- set up vault and use boundary controller
- create a vault credential library
- you can add/update/delete credential mapping overrides

:technologist: [Admin preview](PATH_TO_FEATURE)

:desktop_computer: [Desktop preview](PATH_TO_FEATURE)

## Checklist:

- [x] I have added before and after screenshots for UI changes
- [x] I have added JSON response output for API changes
- [x] I have added steps to reproduce and test for bug fixes in the description
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: ui/admin/app/components/form/credential-library/vault-generic/index.js [hunk 1]
```diff
@@ -30,4 +30,23 @@ export default class FormCredentialLibraryVaultGenericComponent extends Componen
...
+
+  /**
+   * Clear the previously selected key value pair when toggling between credential types on a new form
+   *
+   */
+  @action
+  selectCredentialType({ target: { value } }) {
+    if (
+      this.args.model.isNew &&
+      Object.keys(this.args.model.credential_mapping_overrides || {}).length
```
Comment (ZedLi) [inline ui/admin/app/components/form/credential-library/vault-generic/index.js:46]: Personal taste but I prefer `??` instead these days for clarity

File: addons/api/addon/serializers/credential-library.js [hunk 2]
```diff
@@ -31,9 +32,39 @@ export default class CredentialLibrarySerializer extends ApplicationSerializer {
...
+    const { credential_type, type, credential_mapping_overrides, isNew } =
+      snapshot.record;
+    if (
+      type === TYPE_CREDENTIAL_LIBRARY_VAULT_GENERIC &&
+      credential_mapping_overrides &&
+      credential_type &&
+      !isNew
+    ) {
+      //API expects to send null to fields if it is undefined or deleted
+      const remainingAttrs = Object.entries(
```
Comment (ZedLi) [inline addons/api/addon/serializers/credential-library.js:46]: Why is this needed?

File: ui/admin/app/components/form/field/list-wrapper/select-text/index.hbs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            <Hds::Form::Select::Base
+              @value={{this.newOptionKey}}
+              @width={{@width}}
+              disabled={{@disabled}}
+              aria-labelledby={{keyHeaderID}}
+              {{on 'change' (set-from-event this 'newOptionKey')}}
+              as |F|
+            >
+              <F.Options>
+                <option value='' />
```
Comment (ZedLi) [inline ui/admin/app/components/form/field/list-wrapper/select-text/index.hbs:77]: Shouldn't this be not selectable?

File: addons/api/addon/serializers/credential-library.js (hunk 2 above)
Comment (DhariniJeeva) [reply addons/api/addon/serializers/credential-library.js:46]: When updating the select list, the API explicitly wants us to send null to fields we don't need. Otherwise, they will return the unmodified value from previous versions. 

For ex, If I have a username attr and want to change that to a `password` attr, I will need to send both the fields and make the username null. Otherwise, they will send values for both fields.

File: ui/admin/app/components/form/field/list-wrapper/select-text/index.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+   */
+  get options() {
+    return this.args?.options || this.args?.model?.[this.args.name];
+  }
+
+  /**
+   * Determines if we need to show an empty row to the users to enter more key/value pairs
+   * @type {object}
+   */
+  get showNewRow() {
```
[the remark this answers is not in the document]
Comment (DhariniJeeva) [reply ui/admin/app/components/form/field/list-wrapper/select-text/index.js:36]: actually I had that thought and that would definitely eliminate the need for an extra function but wasn't sure if that was a better option. I can update it

File: addons/api/addon/serializers/credential-library.js (hunk 2 above)
Comment (ZedLi) [reply addons/api/addon/serializers/credential-library.js:46]: I had a setup where I changed the key and it wasn't setting null, just not sending the field which was working fine, is this expected?

https://github.com/hashicorp/boundary-ui/assets/5783847/e53f8d24-fca3-48b3-8268-6d07eebd7b38

File: addons/api/addon/serializers/credential-library.js [hunk 5]
```diff
@@ -31,9 +32,39 @@ export default class CredentialLibrarySerializer extends ApplicationSerializer {
...
+    const value = super.serializeAttribute(...arguments);
+    const { credential_type, type, credential_mapping_overrides, isNew } =
+      snapshot.record;
+    if (
+      type === TYPE_CREDENTIAL_LIBRARY_VAULT_GENERIC &&
+      credential_mapping_overrides &&
+      credential_type &&
+      !isNew
+    ) {
+      //API expects to send null to fields if it is undefined or deleted
```
Comment (ZedLi) [inline addons/api/addon/serializers/credential-library.js:45]:
```suggestion
      // API expects to send null to fields if it is undefined or deleted
```

Can we add a space between the slashes for comments 😅

Comment (DhariniJeeva) [reply addons/api/addon/serializers/credential-library.js:45]: oops I somehow always forget to add spaces there 😿

File: addons/api/addon/serializers/credential-library.js (hunk 2 above)
Comment (DhariniJeeva) [reply addons/api/addon/serializers/credential-library.js:46]: It is set in the serializer, so I don't think you would see it here?   I can see the null being sent to the API in the network call

Comment (ZedLi) [reply addons/api/addon/serializers/credential-library.js:46]: I think I understand now, I don't see it still but it doesn't matter from the API perspective. This was just ember re-using old values bc it assumes they're still relevant if we don't tell it otherwise

File: addons/api/addon/serializers/credential-library.js [hunk 6]
```diff
@@ -31,9 +32,39 @@ export default class CredentialLibrarySerializer extends ApplicationSerializer {
...
+            obj,
+            credential_mapping_overrides[key] ? {} : { [key]: null },
+          ),
+        {},
+      );
+
+      json.credential_mapping_overrides = {
+        ...credential_mapping_overrides,
+        ...remainingAttrs,
+      };
```
Comment (ZedLi) [inline addons/api/addon/serializers/credential-library.js:60]: Few comments regarding this, I don't think you needed to use `Object.entries` since you only used the keys anyways. And couldn't we have included the actual values of the fields as part of the reduce rather than keeping them separate and doing something like this:

```suggestion
      json.credential_mapping_overrides = Object.keys(
        options.mapping_overrides[credential_type],
      ).reduce((obj, key) => {
        if (credential_mapping_overrides[key]) {
          obj[key] = credential_mapping_overrides[key];
        } else {
          obj[key] = null;
        }
        return obj;
      }, {});
```

File: ui/admin/app/components/form/credential-library/vault-generic/index.js [hunk 7]
```diff
@@ -30,4 +30,23 @@ export default class FormCredentialLibraryVaultGenericComponent extends Componen
...
+    return options.mapping_overrides[this.args.model.credential_type];
+  }
+
+  /**
+   * Clear the previously selected key value pair when toggling between credential types on a new form
+   *
+   */
+  @action
+  selectCredentialType({ target: { value } }) {
+    if (
```
Comment (ZedLi) [inline ui/admin/app/components/form/credential-library/vault-generic/index.js:44]: Actually do you need this if check? Couldn't you just set it regardless if they're switching types? It should always be an empty object anyways right?

Comment (DhariniJeeva) [reply ui/admin/app/components/form/credential-library/vault-generic/index.js:44]: good call, yes, I don't need it, since a user can't update this type once its created, I will remove

File: ui/admin/app/components/form/field/list-wrapper/select-text/index.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+      const field = this.args.name;
+
+      const newObj = Object.keys(this.args.model[field] || {}).reduce(
+        (obj, key) => {
+          key !== selectedKey && (obj[key] = this.args.model[field][key]);
+          return obj;
+        },
+        {},
+      );
+      set(this.args.model, field, newObj);
```
Comment (ZedLi) [inline ui/admin/app/components/form/field/list-wrapper/select-text/index.js:120]: This seems unnecessary to reduce over the object when you really just want to delete a key (I also want to mention this component assumes an object now instead of an array, which is unexpected).

You could have done something like: 
```suggestion
      const { [selectedKey]: _, ...newObj } = this.args.model[field];
      set(this.args.model, field, newObj);
```
or

```suggestion
      const newObj = { ...this.args.model[field] };
      delete newObj[selectedKey];
      set(this.args.model, field, newObj);

```

File: ui/admin/app/components/form/field/list-wrapper/select-text/index.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        });
+      }
+    } else {
+      const field = this.args.name;
+      const newRow = {};
+      newRow[this.newOptionKey] = this.newOptionValue;
+      this.args.model[field] = {
+        ...this.args.model[field],
+        ...newRow,
+      };
```
Comment (ZedLi) [inline ui/admin/app/components/form/field/list-wrapper/select-text/index.js:94]: This could also be simplified a bit:

```suggestion
      this.args.model[field] = {
        ...this.args.model[field],
        [this.newOptionKey]: this.newOptionValue,
      };
```

File: ui/admin/app/components/form/field/list-wrapper/select-text/index.hbs (hunk 3 above)
Comment (DhariniJeeva) [reply ui/admin/app/components/form/field/list-wrapper/select-text/index.hbs:77]: I added the text, but if we make this disabled, the select option isn't getting selected when there's one item in the list anyway. The user won't be able to select this empty option and click `add`, since we only enable `add` when there a valid option is selected from the dropdown. I updated the PR description to show this behavior.

File: ui/admin/app/components/form/field/list-wrapper/select-text/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (DhariniJeeva) [reply ui/admin/app/components/form/field/list-wrapper/select-text/index.js:37]: I added a new arg `removeDuplicates` which allows the consumer to decide if they want to use that feature, let me know your thoughts

File: ui/admin/app/components/form/field/list-wrapper/select-text/index.js (hunk 8 above)
Comment (DhariniJeeva) [reply ui/admin/app/components/form/field/list-wrapper/select-text/index.js]: simplified

File: ui/admin/app/components/form/field/list-wrapper/select-text/index.js
[the remark this answers is not in the document]
Comment (DhariniJeeva) [reply ui/admin/app/components/form/field/list-wrapper/select-text/index.js:1]: I thought about that, but this form is essentially a key/value that doesn't allow duplicates and I liked that we didn't have to use our custom transform object-as-array for this..

[the remark this answers is not in the document]
Comment (ZedLi) [reply ui/admin/app/components/form/field/list-wrapper/select-text/index.js:1]: We can probably talk about this component after we get it in but there's no reason this form _shouldn't_ allow duplicates, that happens to be an implementation detail for specifically the vault credential overrides mapping which we've now baked and coupled into this component

File: ui/admin/app/components/form/field/list-wrapper/select-text/index.hbs (hunk 3 above)
Comment (ZedLi) [reply ui/admin/app/components/form/field/list-wrapper/select-text/index.hbs]: Oh but i meant the user should never be able to select this option right? It's not like `credential_type` where the user might want to select it as that "removes" the choice. In this case, the user would just delete the row

File: ui/admin/app/components/form/credential-library/vault-generic/index.hbs [hunk 10]
```diff
@@ -140,6 +140,46 @@
...
+        {{/if}}
+      </:fieldset>
+
+      <:field as |F|>
+        <F.SelectText
+          @name='credential_mapping_overrides'
+          @selectOptions={{this.mappingOverrides}}
+          @model={{@model}}
+          @disabled={{form.disabled}}
+          @removeDuplicates='true'
```
Comment (ZedLi) [inline ui/admin/app/components/form/credential-library/vault-generic/index.hbs:176]: Shouldn't this be a boolean and not a string?

File: ui/admin/app/components/form/field/list-wrapper/select-text/index.hbs (hunk 3 above)
Comment (DhariniJeeva) [reply ui/admin/app/components/form/field/list-wrapper/select-text/index.hbs]: ty, done
